### PR TITLE
make CoreDNS DoH Server

### PR DESCRIPTION
**Pull request by miekg**
### 1. What does this pull request do?
Add experimental DoH support.

### CODE REVIEW

**Comment by miekg** [comment]
Also need a way to for http/2

and the *forward* and maybe *proxy* plugin should start speaking it.

File: core/dnsserver/doh/https.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	"io/ioutil"
+	"net/http"
+
+	"github.com/miekg/dns"
+)
+
+// MimeType is the DoH mimetype that should be used.
+const MimeType = "application/dns-udpwireformat"
+
+// RequestToMsg extra the dns message from the request body.
```
Comment (chantra) [inline core/dnsserver/doh/https.go]: typo: s/extra/extract/

File: core/dnsserver/doh/https.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+func RequestToMsg(req *http.Request) (*dns.Msg, error) {
+	defer req.Body.Close()
+
+	buf, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, nil
```
Comment (chantra) [inline core/dnsserver/doh/https.go]: shouldn't this return the error?

File: core/dnsserver/server_https.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+	s.ServeDNS(context.Background(), dw, msg)
+
+	buf, err := dw.Msg.Pack()
+	if err != nil {
+		// new message?
+	}
+
+	w.Header().Set("Content-Type", doh.MimeType)
+	w.Header().Set("Cache-Control", "max-age=0")
```
Comment (chantra) [inline core/dnsserver/server_https.go]: maybe a FIXME to implement cache-control later?

File: core/dnsserver/doh/https.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	"net/http"
+
+	"github.com/miekg/dns"
+)
+
+// MimeType is the DoH mimetype that should be used.
+const MimeType = "application/dns-udpwireformat"
+
+// RequestToMsg extra the dns message from the request body.
+func RequestToMsg(req *http.Request) (*dns.Msg, error) {
```
Comment (chantra) [inline core/dnsserver/doh/https.go]: note that this only handle POST.
For GET use case, the content-type should be taken from the query parameter `ct` and the dns payload will be base64safeurl encoded (with trailing `=` removed) in the `dns` query parameter. Maybe worth a FIXME/TODO comment.

**Review by chantra — COMMENTED** [review]
Nice!

**Comment by miekg** [comment]
@chantra thanks for reviewing. yes error handling needs to be done - the draft is also light on this. I'll fix things after -04 is released.

**Comment by miekg** [comment]
@chantra PTAL, address all comments and implemented -08 of the draft.

File: core/dnsserver/https.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+// getRequestToMsg extract the dns message from the GET request.
+func getRequestToMsg(req *http.Request) (*dns.Msg, error) {
+	values := req.URL.Query()
+	b64, ok := values["dns"]
+	if !ok {
+		return nil, fmt.Errorf("no 'dns' query parameter found")
+	}
+	if len(b64) != 1 {
+		return nil, fmt.Errorf("multipe 'dns' query values found")
```
Comment (johnbelamaric) [inline core/dnsserver/https.go]: nit: s/multipe/multiple/

File: core/dnsserver/server-https.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	dw.Raddr = &net.TCPAddr{IP: ip, Port: po}
+
+	// We just call the normal chain handler - all error handling is done there.
+	// We should expect a packet to be returned that we can send to the client.
+	s.ServeDNS(context.Background(), dw, msg)
+
+	buf, _ := dw.Msg.Pack()
+
+	w.Header().Set("Content-Type", mimeTypeDOH)
+	w.Header().Set("Cache-Control", "max-age=128") // Minttl as done in cache.
```
Comment (johnbelamaric) [inline core/dnsserver/server-https.go]: No control over this?

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
minor issues, looks good

File: core/dnsserver/https.go (hunk 5 above)
Comment (miekg) [reply core/dnsserver/https.go]: doine

File: core/dnsserver/server-https.go (hunk 6 above)
Comment (miekg) [reply core/dnsserver/server-https.go]: proper bug filed: #1823
The draft is still debating the exact requirements here, but it is clear what needs to be done. Left TODO with issue number.

**Review by johnbelamaric — APPROVED** [review]

**Comment by chantra** [comment]
Great stuff!

**Pull request merged**